### PR TITLE
User/matt.gardner/unittests

### PR DESCRIPTION
**Pull request by matt-gardner2**
Adding an automated test project and put a multi-generational test into it.

### CODE REVIEW

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.csproj [hunk 1]
```diff
@@ -40,6 +40,9 @@
     <Reference Include="Catalyst.DataProcessing.Shared.Utilities, Version=4.4.0.0, Culture=neutral, processorArchitecture=MSIL">
       <HintPath>..\packages\Catalyst.DataProcessing.Shared.Utilities.4.4.1812.1201\lib\net461\Catalyst.DataProcessing.Shared.Utilities.dll</HintPath>
     </Reference>
+    <Reference Include="Catalyst.Platform.CommonExtensions, Version=1.0.0.0, Culture=neutral, PublicKeyToken=e99707dc63721e10, processorArchitecture=MSIL">
```
Comment (mhollier) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.csproj:43]: We are trying to move away from using Catalyst.Platform.CommonExtensions. This functionality has been migrated to Catalyst.DataProcessing.Shared.Utilities.

Comment (matt-gardner2) [reply HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.csproj:43]: Yes.  That's what Jerry said.  We found out that there is a dependency on it, though, for now.

Comment (justin-page) [reply HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.csproj:43]: You need to use the Catalyst.Platform.CommonExtensions for it to work in 18.2

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/HierarchicalDataTransformer.cs [hunk 2]
```diff
@@ -252,7 +252,8 @@ private void ValidateHierarchicalBinding(Binding binding, Binding[] allBindings)
             bool isFirst)
         {
             var sourceEntity = await this.GetEntityFromBinding(rootBinding);
-            LoggingHelper2.Debug($"GenerateDataSources -- sourceEntity: {JsonConvert.SerializeObject(sourceEntity)}");
+
+            // LoggingHelper2.Debug(this.guid, $"GenerateDataSources -- sourceEntity: {JsonConvert.SerializeObject(sourceEntity)}");
```
Comment (justin-page) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/HierarchicalDataTransformer.cs:256]: Should we just delete these debug lines? or are we planning to restore them later?

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+    using Fabric.Databus.Config;
+
+    using Moq;
+
+    [TestClass]
+    public class GrandparentParentChildTest
+    {
+        [TestMethod]
+        public void TestMethod1()
```
Comment (justin-page) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs]: better name for the test?

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+            };
+        }
+
+        private int DataMartId
+        {
+            get
+            {
+                return 123;
+            }
+        }
```
Comment (justin-page) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs]: you can do "private const int DataMartId = 123;" or "private int DataMartId => 123;" if you want to be more concise

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+        {
+            Binding[] bindings = new Binding[3]
+                                     {
+                                         this.GetNestedBindingLevel0Source(),
+                                         this.GetNestedBindingLevel1Source(),
+                                         this.GetNestedBindingLevel2Source()
+                                     };
+            Field[] entity0Fields = this.GetEntity0Fields();
+            Field[] entity1Fields = this.GetEntity1Fields();
+            Field[] entity2Fields = this.GetEntity2Fields();
```
Comment (justin-page) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs:38]: if you're planning on having other tests using these same shared methods, you might want to consider putting them in a TestInitialize method. The mocks may be good to put there too.

**Review by justin-page — APPROVED** [review]

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+                                                               = "Cardinality",
+                                                           AttributeValue
+                                                               = "SingleObject"
+                                                       },
+                                                   new
+                                                   ObjectAttributeValue
+                                                       {
+                                                           AttributeName
+                                                               = "ParentKeyFields",
+                                                           AttributeValue
```
Comment (jmwestbe) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs:227]: I don't see these columns in any of the GetLevelXXFields() methods.  I think that the all three Source entities need to have 'id0' and both level1 and level2 need to have 'id1' in order for all of these relationships to make sense.

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs (hunk 3 above)
Comment (matt-gardner2) [reply HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs]: That's weird - I thought I had renamed it.  I'll make sure it's checked in.

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs (hunk 4 above)
Comment (matt-gardner2) [reply HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs]: Good idea.

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+
+        private Entity GetNestedDestinationEntity()
+        {
+            return new Entity
+                       {
+                           Id = 6357,
+                           DataMartId = this.DataMartId,
+                           Fields =
+                               {
+                                   new Field
```
Comment (jmwestbe) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs:379]: This one also needs to have 'Level0Entity__id0', 'Level1Entity__id0', 'Level2Entity__id0', 'Level1Entity__id1', 'Level2Entity__id1' -- according the parent/child key field designations above.

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs (hunk 5 above)
Comment (matt-gardner2) [reply HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs:38]: Good idea.  For now I'll leave them the way they are and refactor when needed.

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+        }
+
+        private Entity GetLevel0SourceEntity()
+        {
+            return new Entity
+                       {
+                           Id = 0,
+                           EntityName = "Level0Entity",
+                           DataMartId = this.DataMartId,
+                           Fields = { new Field { IsPrimaryKey = true, FieldName = "Level0EntityPrimaryKey" } },
```
Comment (jmwestbe) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs]: Should this (and 406, 419) be calling the GetLevelXFields() methods?

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/HierarchicalDataTransformer.cs (hunk 2 above)
Comment (jmwestbe) [reply HierarchicalDataConverter/Catalyst.HierarchicalDataConverter/HierarchicalDataTransformer.cs:256]: They will all get cleaned up today-ish when we tackle the logging pr.  Looks like we've been going back and forth deleting some and adding some back in between our prs.  Probably okay to reconcile it all together in the logging pr

File: HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+            serviceClientMock.Setup(mock => mock.GetEntityAsync(1)).Returns(Task.FromResult(this.GetLevel1SourceEntity()));
+            serviceClientMock.Setup(mock => mock.GetEntityAsync(2)).Returns(Task.FromResult(this.GetLevel2SourceEntity()));
+
+            var converter = new HierarchicalDataTransformer(serviceClientMock.Object);
+            var privateMethodRunner = new PrivateObject(converter);
+            object[] args = new object[2] { this.GetNestedBindingLevel0Source(), this.GetNestedDestinationEntity() };
+
+            var jobData = ((Task<JobData>)privateMethodRunner.Invoke("GetJobData", args)).Result;
+
+            Assert.AreEqual(3, jobData.DataSources.Count());
```
Comment (jmwestbe) [inline HierarchicalDataConverter/Catalyst.HierarchicalDataConverter.AutomatedTests/GrandparentParentChildTest.cs:57]: These assertions seem to be missing the Relationship validations

**Review by mhollier — APPROVED** [review]

**Pull request merged**